core/src/main/java/org/aerogear/mobile/core/http/OkHttpResponse.java

```diff
@@ -23,12 +25,18 @@ public OkHttpResponse(final Call okHttpCall, AppExecutors appExecutors) {
             try {
                 response = okHttpCall.execute();
                 requestCompleteLatch.countDown();
-                if (completionHandler != null) {
-                    runCompletionHandler();
+
+                if (!response.isSuccessful()) {
+                    throw new IOException("Request failed with status code " + response.code());
```

Comment:
I'm in doubt what we should do here. The _request_ was **success** called, but the response was not (range [200..300)). This should be a dev responsibility to handle that. IMO we should always call the success expect when the HTTP layer raises a problem. That is how retrofit handles it.

@secondsun Can you give you opinion here?

Reply:
So first a bit of bikeshedding, IOException is the wrong exception to use here.  IOException is more of a "the network quit working" type of failure.

I think that success and error handlers make a lot more sense when we start thinking about ways for modules to "rescue" a http request.  If auth can refresh a token then it should be able to register as an error handler and then this distinction makes a lot of sense.

However this kind of changes the semantics a bit.  A request is made.  A response is returned.  success and error handlers are called (which may or may not generate other network activity as appropriate) and when the whole mess has settled the completion handler has finally been called.  Completion therefore stops being about a single HTTP request and is instead about a user's action in the application.

Sorry I don't think I made anything more clear LOL